### ISSUE TITLE
Create Actions for Review of accuracy of Civil Parishes in FC1 vld files

### ISSUE THREAD

**Issue by geoffj-FUG** [FreeUKGen/FreeCENMigration#1397]
We now test the civil parish in csv files against the PARMS to ensure that they are entered correctly. 
As far as I am aware the old vld files loaded into FC1 did not have this check. In addition the FC2 PARMS was updated during the cutover to the new system.
The work that was done linking the PARMS has made our public searches more accurate. However the historic data has not had the alignment done.
Can we get a report on the civil parishes that do not align with the PARMS please? 
We will need the piece number, ED number and the civil parish that has been recorded but dos not appear in the PARMS against that piece. We also need the PARMS civil parishes for that piece. We only need a report entry once for every unique civil parish in an ED within a piece where there is not a match.
Once we know the extent of the problem then we can take action to update the historic data and that will make our existing searches more accurate.
I am not sure what the impact of breaking down pieces into parts in FC2 will have on this report. We need to have a look at this as well and this report will show us whether there has been any.
Geoff

**Comment by Captainkirkdawson** [comment]
@geoffj-FUG could you clearly identify the purpose of this activity. Why do we need to do it?
I know you say "make our existing searches more accurate" but I fail to see how.
@PatReynolds for information
You may recall that in the latter part of last year we made the transition to use FreeCEN2 Places for both input streams instead of Places (freecen1) for vld files and FreeCEN2 Places for csv. This ensured that we used a consistent place name **and its location**. The latter enabled the reintroduction of the Nearby Places option. Personally I am not convinced that neither the team nor the research community realise the significance of this transition.

As we say in the Help for researchers
**A Place is usually the location associated with the Registration Sub District used by the Census Enumerator: where this is unknown we use the location of the Registration District. With the transcriptions started in 2021 we use the location that is associated with the Civil Parish and which provides a more localised set of records.**

Fundamentally this means that csv records are searched with a greater granularity than vld records.
We achieved this by connecting every vld to its appropriate freecen2_piece and its freecen2_place. All search records for each vld were then updated to include the information for freecen2_place.

As you are all too aware vld files are not segregated by ED or civil parish hence FreeCEN2 has never had a civil parish or ED search capability. 

**This is because there is fundamental difference in the design of FreeCEN2 compared with FreeCEN1.**
FreeCEN1 had 2 search collections (Tables). A Household collection that contained all of the fields for a dwelling (Census	County	District	Civil Parish	Ecclesiastical Parish	  Piece	Enumeration District	Schedule	House Number	House or Street Name) **each of which was indexed** and an Individual collection with each individual and all its fields joined to its household. 

FreeCEN2 only has a single search collection. The search record. This has only a few of the individual fields and a very few household fields(Census	County Place). However for vlds it displays the details from 2 non searchable collections (dwellings and individuals) and for csv from the csv entries collection.

**Bottom line is that the FreecCEN1 search is a much more powerful search engine than the current implementation of FreeCEN2 **

In summary, the Ecclesiastical Parish, ED and Civil Parish information displayed by FreeCEN2 are **informational items** only. 

The enormous amount of work being proposed in this story to reconcile the Ecclesiastical Parish, ED and Civil Parish in vld entries with the FreeCEN2 parameters is only warranted if it is intended to add these as searchable fields in what one would likely want to call FreeCEN3!

**Comment by Captainkirkdawson** [comment]
The first step would be making the civil parish and ecclesiastical parish entries for a file consistent. 
There are 10557 vld files. These have 18160 unique values of civil parishes and 16645 ecclesiastical parish in freecen_dwellings. There was no quality control on these fields so there are variations.

Some examples of the variation in civil parish names as entered.
        "Toxteh Park",
        "Toxteht Park",
        "Toxteth Park",
        "Toxteth Park Lan",
        "Toxtheth Park",

        "Upton Cum Chalvey",
        "Upton Cum Chanley",
        "Upton Cun Chalvey",
        "Upton With Chalvey",

This work alone I estimate to be a 21 .

The second step would be to compare the civil parish names for a vld file with the freecen2_civil_parish names for the freecen2_piece of the vld files. I estimate to be a 21 level of effort

**Comment by geoffj-FUG** [comment]
Kirk

That is exactly why I created this story.

There was no quality control on the entries. Your examples of Toxteth Park show that the problem may be bigger than I anticipated. 

The civil parish was entered by the transcriber regularly throughout an ED, rather than just at the first entry as we now do in FC2. That means that the entry could be mis-typed within the ED.

The ecclesiastical parish is a bigger problem. When WinCC was set up the Ecclesiastical Parish field was completed using references transferred to WinCC from the PARMS. This created a debate when we redeveloped the front end of FreeCEN2 as CSVProc regarding TNA records and the recording of civil or ecclesiastical parishes. It is now well established that the PARMS contains civil parishes not ecclesiastical parishes. The amount of data affected is huge so we must assume that we cannot do anything with it. We would need to go back and check every ED in every piece to correct the problem. So, in my view, the ecclesiastical parish data is a loss. It is not practical to bring it up to scratch and we do not have a reference file to compare it with.

The wish to upgrade the search report capabilities of FC2 to match FC1 has been raised as the next step in the development of FC2. The accuracy of any new searches is dependent on the accuracy of the underlying data. The civil parish data struck me as an area that might be able to be improved given that it is usually limited to one entry per ED, and for those EDs with more than one civil parish it is only a very few. Hence the desire to see how big the problem was by doing a match with the PARMS to identify those civil parishes not in the PARMS. 

I had not expected to see the results in Kirk’s review of the data, given that the process is Transcribe, Proofread and Validate. It shows that the Proofreader has not been picking up these typos. The Validator would not have seen them as they were not tested in Valdrev as they are now by the CSVProc upload.

 We have also made the effort to align the PARMS with the Gazetteer, with a significant loss of Coordinator goodwill in the process. So, one step in the integrity links has already been achieved. 

It strikes me that the need to identify these civil parish variations is now even more important as they affect the usefulness of the collected data. We have introduced the data integrity tests into FC2 so it seems logical to at least examine the extent of this data mismatch problem.

If it is practicable to update the civil parish data then it becomes possible to use this field to search, for instance the civil parish as a sub-set of the census place or maybe even as an optional alternative to the census place.

I am not sure what a 21 work estimate means.  Is there a scale reference that I can look at please?

Geoff

**Comment by Captainkirkdawson** [comment]
@geoffj-FUG efforts are ball park guess of the amount of time an experienced person might take. 1 is a line of code; 3 is an hour or so; 5 up to a day; 8 up to a week; 13 up to a month; 21 up to 4 months; 40 a year or more.

**Comment by geoffj-FUG** [comment]
Thank you

 

Geoff

**Comment by PatReynolds** [comment]
We need to revisit the FreeCEN roadmap, and allocate this to a part of it.

**Comment by geoffj-FUG** [comment]
Pat

 

This is just one story.

 

The Epic is review accuracy of existing data uploaded from FC1. There are several issues involved. All issues are created by vld files having no integrity testing. All csv files uploaded now are integrity tested but vld files uploaded now are not and can’t be.

 

The first thing that we should clean-up is the Civil Parishes. These were repetitively entered in FC1 and not tested. The incidence of typos is significant. (For csv files they are entered once per ED  unless they change within an ED and integrity tested against the PARMS). The other cause of incorrect civil parishes is the need in FC1 to abbreviate some of them to fit the spreadsheet column width (which was governed by Wincc and Valdrev constraints).

At the moment researchers cannot search against the civil parishes within a piece. If they could the search would produce poor results because many civil parishes are mis-spelt. 

Once cleaned up we have the option of introducing searching civil parishes as a subset of the Registration District (or even of the County).

 

The second data field that needs looking at is the Places of Birth. There are a huge number of entries on the database that are not in the Gazetteer. There are a number of causes including (in no particular order) typos, abbreviations, validators adding incorrect spellings to their test file to eliminate repetitive mis-spellings by the enumerators. Places we have not yet added to the Gazetteer, the Gazetteer now using the greater to smaller rule etc. 

No work can start on POB searching until this has been sorted out.

FC1 and FC2 have different validation policies – 

FC1 was if there is evidence that a place reasonably existed, pass it or note that it cannot be located. 

FC2 is if a place is in the Gazetteer pass it or note that it cannot be located. If it is in the Gazetteer it existed, has a url as a reference and as it is geolocated it can be linked to Google Maps and its location displayed. It is this change in policy that will make high quality POB searching possible.

 

Any attempt to move forward on searching at the moment will result in a garbage in, garbage out scenario.

 

So I agree, we need to revisit the roadmap, because at the moment there is a huge hole in the road! 

 

Geoff

**Comment by PatReynolds** [comment]
There is no Epic with the word Review in it (there is only one Epic - the Accessibility Review.  Do you mean _this story_ is an epic? Or should I create an epic and put this and other stories in it?

**Comment by geoffj-FUG** [comment]
Pat

 

There will be a at least two stories related to data integrity. They amount to something quite large.

 

I did not use the term in its true context but it would do no harm to tie them together with an epic. Any other data integrity issues can then be added so that we can see the true size of the problem.

 

Geoff

**Comment by geoffj-FUG** [comment]
This would be part of the clean-up of the vld files. It cannot proceed until the vld uploads are no longer happening.
Geoff

**Comment by geoffj-FUG** [comment]
**Testing of Civil parishes within FC1 vld data collection**

We will need to run a report showing:
Unique (Piece Number, ED, Civil Parish), where Civil parish for that piece is not in the PARMS

This report should be produced first so that we can see the extent of the problem.
The PARMS entry will need to be looked at for each piece to ascertain whether the problem is in the transcription or in the PARMS.
There is no Download function as a vld file. To download we would need to re-validate the piece and that is not desirable. Therefore, we will need to work on the live data in the vld data collection. Before any changes are made, we will need to back up the collection.
Every entry within an ED may have the same incorrect Civil parish, depending on what caused the problem. If it is just one Folio that was affected, it can still affect 40 to 50 records. We therefore need to amend the live data set. 
_Question: The vld data collection is a fixed length flat file. Is it easier to work on it in this format or to download it as a csv file and then replace that vld file with an flat file generated from the csv file?_
An edit routine will need to be established which will allow 
Replace all Civil Parish 
With Civil Parish name
Where Piece number is current piece number and Civil parish is current Civil Parish Name.
Once done we can re-run the original report to ensure that everything is now aligned with the PARMS file.
This should probably be run by the data manager as she has the ability to amend the PARMS where the problem is in that data and not the transcription.
Geoff

**Comment by geoffj-FUG** [comment]
Although this is blocked by story Cease vld Uploads is it possible to develop the report now so that we can see the magnitude of the problem. We can then decide how best to tackle the amendments.

**Comment by geoffj-FUG** [comment]
@AnneV-Learn 
Story #1397 (was in product backlog but I have moved it to In Progress). – this can be commenced prior to the vld data collection being closed as any problems will be one-off for each piece. 
Can we get a report of pieces with Civil Parishes that are not in the PARMS please? We will need the piece number, ED number and the civil parish that has been recorded but does not appear in the PARMS against that piece. Unique (Piece Number, ED, Civil Parish), where Civil parish for that piece is not in the PARMS. 
We only need a report entry once for every unique civil parish in an ED within a piece where there is not a match. We also need the PARMS civil parishes for that piece, but we can look those up in the County options (we must anyway to ascertain where the problem is).
There will be occasions where there is a mismatch between the piece number in the data set and the piece number in the PARMS. So, our report should test for the first characters followed by a underscore and then the text before the next underscore in the PARMS. That will eliminate problems with the PARMS numbering.
This report will be run no more than half a dozen times – once for the first pass, again when the data has been amended, again when we close vld uploads and again when any additional problems have been fixed. (plus a couple for retries!).
If it can be in csv format then the report can be loaded into a spreadsheet and amendments managed with the spreadsheet.
Geoff

**Comment by AnneV-Learn** [comment]
@geoffj-FUG I have coded the rake task to do this and Vino deployed it to Test3 today. I constructed it so that it works through a max of 500 VLD files at a time (although it does work on complete counties, so if it hits the limit of 500 vld files part way through a county it will go on to complete that county but not move onto the next county). This was done so that running it across the Production environment would be manageable. It outputs a log file (summarising the processing) and a CSV file. The file names include the span of chapman codes that it covers. 
In the CSV file you may see **MISSING** in the fc2_piece_number and/or fc2_piece_civil_parishes column, this indicates the link from the VLD file to the fc2 Params is missing, which may occur quite a lot on the Test3 database as to my knowledge the test data there has never been fully updated to the full fc2 structures across the board and certainly does contain anomalies. 
I just ran it once on **Test3** today and I will email you the output files for your review and comment before  progressing any further.

**Comment by geoffj-FUG** [comment]
Anne

That report looks great, thank you.
The Missing links that I have looked at so far are all Scotland. We will need to address this if there is not a Scottish PARMS data available yet. One option would be to copy the FC1 PARMS for Scotland into the new version. However I suspect that something exists because otherwise our searches would not work.
What really stands out in the report is that we need to be able to do a find/replace on all the civil parishes, in order:
Find all '&' and replace with ' and '
Find all full stops and replace with a space
Find all commas and replace with a space
Find all space space and replace with space (to remove double spaces generated by above replacements).

Can we arrange for these find/replacements in the main data set please and thn run the report on the main database.

Can I have an option in the vld files menu to run the report County by County please? It will need to be at Master Coordinator and Data Manager level. Then I can start looking at the necessary changes.

We will also need a passworded find / replace that I can run to fix up civil parish names. It should be limited to changes within a particular piece.

Geoff

**Comment by AnneV-Learn** [comment]
@geoffj-FUG Bear in mind that the report I sent you was from the Test3 data and only covered chapman codes AGY - CAE. 

I’ll start looking at how best to do the Civil Parish find/replace (lots of records to work through there) and adding the option to run the CP not in PARMS report from the Manage VLD page. 

When looking at Civil Parish **names** I see that many of those with chapman code = MIL have commas and ampersands. 
E.g. 
Bermuda, ASG, APG, Perm Staff BORG & A P Depts.
Do you want the commas and ampersands replaced in those?

Also the civil parish **standard_name** has 
hyphens converted to a space, 
occurrences of .,'(){}[] characters removed, 
saint replaced with st 
multiple spaces reduced to one space 
and trailing spaces removed.
Does any of that affect the find/replace you have requested for the Civil Parish **name**?

Also for the listing that I developed (above), I am wondering if I should have been comparing a VLD 'standardised' civil_parish name with the civil parish standard_names for the piece? All very complicated, I hope that makes sense(?)

**Comment by geoffj-FUG** [comment]
@AnneV-Learn 

The characters that I identified are not standardised but create ERROR messages. That is why they need amending.
I made Vino aware at the scrum that I was suggesting a find/replace. You may like to discuss this functionality with her.
She is happy for us to work on the vld data collection itself, but is still thinking about how to achieve it.
In addition we will need a find/replace availabe as we do the corrections. Replace in this insatnce should be restricted to the piece we are working with. i.e. Find civil parish in piece number and replace with new civil parish in the same piece number.
Geoff

**Comment by AnneV-Learn** [comment]
@geoffj-FUG 
I realise that the characters that you identified are not standardised but was wondering regardless of that if I should be comparing standardised place names in the listing of invalid civil parishes in VLD file data.
Yes I will discuss the whole find/replace requirement with @Vino-S when I have given it some more thought myself.

**Comment by geoffj-FUG** [comment]
The special characters advised by Kirk never seem to have been implemented. I have therefore amended the Handbook to show the original text. The priority now is to have an edit function available to me so that I can amend civil parishes within a piece. This should be on the vld dataset and can be implemented as soon as it is available.
Geoff

**Comment by AnneV-Learn** [comment]
@Vino-S would you be able to contact Lemon regarding the issue we discussed last week about Test3 and spawned processes not appearing to work, as I can't release for user testing until that is resolved.

**Comment by Vino-S** [comment]
In discussion with Lemon about the issue

**Comment by DeniseColbert** [comment]
Lemon has fixed spawned process issue, just needs testing that the fix works.

**Comment by AnneV-Learn** [comment]
Thanks @vino-s  - that appears to have fixed the spawned process issue on Test3.

**Comment by AnneV-Learn** [comment]
@geoffj-FUG Just to let you know that I have completed the 'cleaning' of the production VLD file Civil Parish names as per your specification:
_Find all '&' and replace with ' and '
Find all full stops and replace with a space
Find all commas and replace with a space
Find all space space and replace with space._
Obviously if any of the VLD files are re-loaded then those 'issues' may re-appear but that should be too many and can be handled via (the soon to be available) VLD Civil Parish editing functionality.
I hope to have the Listing of Invalid Civil Parishes and Editing of Civil Parishes for VLD files available for testing on Test3 soon.

**Comment by AnneV-Learn** [comment]
@geoffj-FUG The VLD File List Invalid Civil Parishes and Edit Civil Parish functionality is Ready for Testing on Test3. I have sent you an email to explain what is there.

**Comment by geoffj-FUG** [comment]
Anne

 

The spreadsheet looks good.

 

In addition to my query just now, I think that FreeCEN searching uses the text converted to lower case. Look at Row 75 in the spreadsheet for Somerset. Weston super Mare is correct in the PARMS but there is an upper case in the civil parish. 

 

If the system converts to lower case for searching can we allow the Weston Super Mare version through with no impact?

 

There will be several of these instances. If we can ignore the case then can you amend the code to do that please? It will make life so much easier.

 

Geoff

**Comment by geoffj-FUG** [comment]
Anne
The find/replace function works perfectly. 
Can we change it so that it cycles back to the Edit form please? You will see from the Somerset spreadsheet that there may be many entries for the same piece. By cycling back to the edit form I will not have to keep selecting Show for the same piece.
Geoff

**Comment by AnneV-Learn** [comment]
@geoffj-FUG - CSVProc validation of Civil Parish does indeed match the lowercase version of the name, so I'll make that change to the Listing of Invalid Civil Parishes.
I'll also change so that it returns to the Edit page (rather than the index) after a successful edit.
I'll let you know when Test3 code has been updated so that you can then re-test.

**Comment by AnneV-Learn** [comment]
Ok this changes have now been deployed to Test3, so ready for you to test @geoffj-FUG 
Could you also confirm that system_administrator, data_manager and master_county_cooordinator are the correct roles to have access to the new List Invalid and Edit Civil Parishes functionality.

**Comment by geoffj-FUG** [comment]
Anne

 

Can you make it available to Coordinators.

 

Although I will not be asking coordinators to do it, I know that there are a couple who will.

 

I will have a look at it now.

 

Geoff

**Comment by geoffj-FUG** [comment]
Anne
That is all very good. I have now come across another problem. If you look at RG511923 there are a whole series of entries for Bishops Lydeard (same for North Petherton in the next piece). You will see that there are lots of options in the PARMS listing. So I need to be able to choose whether I am amending the whole piece or just the ED. Otherwise I will be changing every entry in the piece and not be able to see what has been changed (in other words I can make a mess).
That was an unforeseen problem.
Can you put this option in please? 
Maybe we can put a list box with the EDs that the selected entry occurs in (or even with all the ED numbers) so that the ED can be selected and the change only happen in that ED. Make the first option All EDs. The submit button would be below this box.
Geoff

**Comment by geoffj-FUG** [comment]
Anne
It has just struck me that another option is to list the ED next to the civil parish in the dropdown listing of the edit form. Then that can be selected and the change made to that ED alone. EDs will need to be looked at individually when these instances occur so it will not make a big difference to the work involved fixing the entries up.
Geoff

**Comment by AnneV-Learn** [comment]
@geoffj-FUG Yes I did wonder if that might be acceptable. Keeps the form simple too. I’ll re-code for that and then you can re-test to assess if it is practical. Thinking the drop-down would show ED followed by ‘ - ‘ then the Civil Parish eg ‘1 - Farncombe’, is that ok or would you prefer a different format?

**Comment by geoffj-FUG** [comment]
Anne

Yes. 1 - Farncombe format will be good.

Geoff

**Comment by AnneV-Learn** [comment]
@geoffj-FUG Test3 has now been updated as discussed above. 
Well slight change - the  dropdown shows ED followed by ‘ -- ‘ then the Civil Parish eg ‘1 -- Farncombe’ so has to clearly differentiate from Civil Parishes names that could contain single hyphens.
So over to you for re-testing.

**Comment by geoffj-FUG** [comment]
See story #1493 this additional action at data manager level will be needed to clean up the civil parish entries. The Scotland Coordinator will also need this new action.
Anne, How big a task is it please?

**Comment by geoffj-FUG** [comment]
This looks good and works well. I have marked it as ready for deployment. I have created story #1497 to enable an action that we will need to complete this task.
Geoff

**Comment by geoffj-FUG** [comment]
I have thought about permissions. Yes, make it available to coordinators as well please
Geoff

**Comment by geoffj-FUG** [comment]
Works well in live system. However I have found a new issue.
If we have a look at the Somerset listing we have HO107944. It is showing a mismatch. In the PARMS there are two entrie that make up the piece - HO107_944A and HO107_944B. The entry has only been tested by the routine against 944A. It has rejected the entries in 944B. So we need some way of making sure that we test against all the PARMS entries. 
This has occurred because the vld data is not split into the new pieces but is one intact piece. 
I am ignoring the correct entries at the moment but they will still report.
Geoff

**Comment by AnneV-Learn** [comment]
@geoffj-FUG OK I’ll look into that. I didn’t realise that one VLD file could be linked to more than one PARMS record.

**Comment by AnneV-Learn** [comment]
@geoffj-FUG I’ve updated the code for the report and it is now available for testing on Test3.

**Comment by geoffj-FUG** [comment]
I don't have suitable data in the test3 environment. Thats why it was missed during test last time. 
Deploy it and I will check it there. I am the only user so the risk is very low.
Geoff

**Comment by AnneV-Learn** [comment]
@geoffj-FUG If I have understood the issue correctly I think SOM HO511925.VLD might be a case you could check on Test3.

**Comment by geoffj-FUG** [comment]
Another request please. Can we ignore hyphens in the PARMS please and treat them as a space? (if that is the way that the system matches in the csv data collection tests). If the hyphens cause a mismatch on the main system I will need to take a different approach with the matching. I have worked through about 17- changes and am beginning to pick up these little quirks.
Geoff

**Comment by geoffj-FUG** [comment]
Anne

Looks good. Fingers crossed!

Geoff

**Comment by DeniseColbert** [comment]
Deployed: closing
